oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,213 @@
+| Status          | Draft                                                  |
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2018-03-27                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | <start - target end dates for review>                  |
++-----------------+--------------------------------------------------------+
+| `Resolution`    | <links to any discussions where the final              |
+|                 | status was decided>                                    |
++-----------------+--------------------------------------------------------+
+
+.. contents::
+   :local:
+   :depth: 2
+
+Context
+-------
+
+Developers who work on the Open edX platform, both inside and outside of edX,
+need accurate and current documentation about platform architecture, APIs, and
+development best practices.
+
+Developer documentation has been inconsistently and in some cases poorly
+maintained. Documents exist on ReadTheDocs (RTD), in repositories on GitHub,
+in Confluence, and in Google Docs. Documents are also difficult to find.
+
+This proposal focuses on long-term developer "Documentation" needs and does not
+concern itself with any transitory “Notes” that a development team may use for
+project management, brainstorming, meeting notes, etc.  We also distinguish
+"Documentation" from any useful code "Comments" that a developer may choose
+to add to clarify or add rationale to a local code segment. "Comments" are also
+out of scope of this OEP.
+
+Decisions
+---------
+
+.. image:: oep-0019/developer_docs.png
+   :alt: A diagram that shows the different types of documentation and their
+    relationships.  This is further described below.
+
+Agile Documentation Needs
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With Agile development and Clean Code, we can minimize over-documentation since
+code should be readable and self-documenting. Where we do still need developer
+documentation is:
+
+1. **Decision Records** to understand the rationale behind the code and the
+   evolution of its perceived tech debt.
+
+   * These will be captured in either: (1) the `open-edx-proposals repo`_ for
+     system-wide decisions or (2) in `local decisions`_ folders within local code
+     repos. The latter may exist in a top-level folder in a repo and also as a
+     subfolder of an individual Django app.
+
+2. **Diagrams** to provide a visual holistic perspective of the overall system 
+   or component with its boundaries, interconnections, and relationships.
+
+   * These will be stored in Github co-located with the relevant code or text.
+
+3. **How-To Documentation** with step-by-step instruction in enabling and using
+   a feature.
```

Comment:
How do we distinguish this from the user documentation about enabling and using a feature?

Reply:
@nedbat Can you take a look at the latest version of the OEP and suggest any further clarification if needed?